File: graphs & tries/841.py
# Google
# Tc = O(n+m)
# Sc = O(n+m)

class Solution:
    def dfs(self, rooms, source, visited):
        visited[source] = True
        for node in rooms[source]:
            if not visited[node]:
                self.dfs(rooms, node, visited)
    def canVisitAllRooms(self, rooms):
        n = len(rooms)
        visited = [False] * n
        self.dfs(rooms, 0, visited)
        for x in visited:
            if not x:
                return False
        return True
         

File: graphs & tries/841.js
class Solution {
    dfs(rooms, source, visited){
        visited[source] = true
        for(let node of rooms[source]){
            if(!visited[node]){
                this.dfs(rooms, node, visited)
            }
        }
    }
    canVisitAllRooms(rooms){
        const n = rooms.length
        const visited = new Array(n).fill(false)
        this.dfs(rooms, 0, visited)
        for(let x of visited){
            if(!x){
                return false
            }
        }
        return true
    }
}